tools/keepalived/keepalived/libipvs-2.6/libipvs.c

```diff
@@ -691,45 +752,37 @@ void ipvs_sort_dests(struct ip_vs_get_dests *d, ipvs_dest_cmp_t f)
 ipvs_service_entry_t *
 ipvs_get_service(struct ip_vs_service_user *hint)
 {
-	ipvs_service_entry_t *svc,*svc_rcv;
+	ipvs_service_entry_t *svc;
+	struct dp_vs_service_entry dpvs_svc, *dpvs_svc_ptr, *dpvs_svc_rcv;
 	socklen_t len;
 	size_t len_rcv;
 
-	ipvs_func = ipvs_get_service;
-
 	len = sizeof(*svc);
 	svc = calloc(1, len);
 	if (!svc)
 		return NULL;
-	len_rcv = len;
 	memset((void *)svc, 0x00, len);
 
-	svc->fwmark = hint->fwmark;
-	svc->af = hint->af;
-	svc->protocol = hint->protocol;
-	svc->addr = hint->addr;
-	svc->port = hint->port;
-	snprintf(svc->srange, sizeof(svc->srange), "%s", hint->srange);
-	snprintf(svc->drange, sizeof(svc->drange), "%s", hint->drange);
-	snprintf(svc->iifname, sizeof(svc->iifname), "%s", hint->iifname);
-	snprintf(svc->oifname, sizeof(svc->oifname), "%s", hint->oifname);
-
-	CHECK_COMPAT_SVC(svc, NULL);
+	len = sizeof(dpvs_svc);
+	len_rcv = sizeof(*dpvs_svc_rcv);
+	memset(&dpvs_svc, 0, len);
+	dpvs_svc_ptr = &dpvs_svc;
+	IPVS_2_DPVS(dpvs_svc_ptr, hint);
+
 	if (dpvs_getsockopt(DPVS_SO_GET_SERVICE,
-		       svc, len, (void **)&svc_rcv, &len_rcv)) {
-		free(svc);
-		dpvs_sockopt_msg_free(svc_rcv);
-		return NULL;
+		       &dpvs_svc, len, (void **)&dpvs_svc_rcv, &len_rcv)) {
+		goto out_err;
+	}
+
+	DPVS_2_IPVS(svc, dpvs_svc_rcv)
+	if (svc->af == AF_INET) {
+		svc->pe_name[0] = '\0';
+		svc->__addr_v4 = svc->addr.ip;
```

Comment:
`svc->__addr_v4 = dpvs_svc_rcv->addr.ip;` ?

Reply:
svc.addr is union struct which can be ipv4/ipv6; ipvasdm/keepalived always read __addr_v4 when protocol is ipv4. So just override here.